src/ansiblelint/transformer.py

```diff
@@ -0,0 +1,253 @@
+        yaml.explicit_end = self.yaml_explicit_end  # type: ignore[assignment]
+        yaml.width = self.yaml_width  # type: ignore[assignment]
+
+        yaml.default_flow_style = False
+        yaml.compact_seq_seq = (  # dash after dash
+            True  # type: ignore[assignment]
+        )
+        yaml.compact_seq_map = (  # key after dash
+            True  # type: ignore[assignment]
+        )
+        # Do not use yaml.indent() as it obscures the purpose of these vars:
+        yaml.map_indent = 2  # type: ignore[assignment]
+        yaml.sequence_indent = 4 if self.yaml_indent_sequences else 2  # type: ignore[assignment]
+        yaml.sequence_dash_offset = yaml.sequence_indent - 2  # type: ignore[operator]
+
+        if self.yaml_indent_sequences:  # in the future: or other formatting options
+            # For root-level sequences, FormattedEmitter overrides sequence_indent
+            # and sequence_dash_offset to prevent root-level indents.
+            yaml.Emitter = FormattedEmitter
+
+        self._yaml = yaml
+        return yaml
+
+    @staticmethod
+    def _final_yaml_transform(text: str) -> str:
+        """Transform YAML string before writing to file.
+
+        Ansible uses PyYAML which only supports YAML 1.1, so we dump YAML 1.1.
+        But we don't want "%YAML 1.1" at the start, so drop that.
+        """
+        prefix = "%YAML 1.1\n"
+        prefix_len = len(prefix)
+        if text.startswith(prefix):
+            return text[prefix_len:]
+        return text
+
+    def run(self) -> None:
+        """For each file, read it, execute fmt transforms on it, then write it."""
+        for file, _ in self.matches_per_file.items():
+            # str() convinces mypy that "text/yaml" is a valid Literal.
+            # Otherwise, it thinks base_kind is one of playbook, meta, tasks, ...
+            file_is_yaml = str(file.base_kind) == "text/yaml"
+
+            try:
+                data: Union[CommentedMap, CommentedSeq, str] = file.content
+            except (UnicodeDecodeError, IsADirectoryError):
+                # we hit a binary file (eg a jar or tar.gz) or a directory
+                data = ""
+                file_is_yaml = False
+
+            if file_is_yaml:
+                # ruamel.yaml only preserves empty (no whitespace) blank lines.
+                # So, drop spaces in whitespace-only lines. ("\n  \n" -> "\n\n")
+                data = _whitespace_only_lines_re.sub("", cast(str, data))
+                # load_data has an lru_cache, so using it should be cached vs using YAML().load() to reload
+                data = load_data(data)
+                with StringIO() as stream:
+                    # ruamel.yaml does not have `dumps`. We have to give it a path or a stream.
+                    self.yaml.dump(data, stream)
+                    file.content = self._final_yaml_transform(stream.getvalue())
+                file.write()
```

Comment:
@lilatomic How's this? Setting the content is now separate from writing the file. What else would you do to this?